update the porting guide (#184)
linuxboot/heads-wiki: Development/Porting.md

nav_order: 3
parent: Development
---

<!-- markdownlint-disable MD033 -->
<details open markdown="block">
  <summary>
    Table of contents
  </summary>
  {: .text-delta }
1. TOC
{:toc}
</details>
<!-- markdownlint-enable MD033 -->

Prerequisites for porting Heads
===
Prerequisites: 
1. coreboot port:  since Heads is a coreboot payload, the board must have fully completed and actively maintained coreboot support. Any known issues should be acceptable to end users. Exceptions include the Purism and Dasharo coreboot forks, where coreboot is used from their respective forks.

> 2. TPM module: Measured boot and the usable security features are at the core of what Heads is about. These features depend on the TPM. Therefore, the board must have a TPM module. If it uses fTPM and you plan to neuter the Intel Management Engine (ME), the fTPM must remain functional under coreboot even after neutering ME. Otherwise, a dTPM is required for this board.

Reviewer: Actually, I think given the size of the Heads payload neutering ME is a prerequisite on most boards. I would make this explicit here (subsequent bullet point numbering might need fixed): 

```suggestion
2. TPM module: Measured boot and the usable security features are at the core of what Heads is about. These features depend on the TPM. Therefore, the board must have a TPM module. If it uses fTPM and you plan to neuter the Intel Management Engine (ME), the fTPM must remain functional under coreboot even after neutering ME. Otherwise, a dTPM is required for this board.
3. The final flash layout must have enough space for the Heads payload. Concretely, the size of the flash chip(s) must be at least be 12MB and the size of the BIOS region must be large enough to take the Heads payload. Usually, this means shrinking another region, the ME, and reallocating the freed space to the BIOS region. However, changing the ME blob or flash layout is not always possible as Intel Bootguard prevents booting with such modifications in most modern boards unless the board vendor chose otherwise.
```

This prerequisite is another reason why it would make sense to move the section on binary blobs even above the board config section.

Author: done

```
CONFIG_COREBOOT_CONFIG=config/coreboot-NEW_BOARD.config
CONFIG_LINUX_CONFIG=config/linux-NEW_BOARD.config
```
* For early testing, enabling debug mode is a good idea. However, these options should be removed before merging:
```
export CONFIG_DEBUG_OUTPUT=y
export CONFIG_ENABLE_FUNCTION_TRACING_OUTPUT=y
```
Note: This may cause glitches where the screen output appears overwritten until an arrow key is pressed. This is normal for DEBUG mode, as the additional tracing output uses the same console as fbwhiptail. The issue will disappear once debug options are removed from the board configs.
* It is important to define the correct TPM configuration (TPM 1.2 vs. TPM 2.0), ensuring they are mutually exclusive, similar to the coreboot configuration (described in coreboot config below). The selected option misses `#`:
```
#TPM2 requirements
export CONFIG_TPM2_TOOLS=y
export CONFIG_PRIMARY_KEY_TYPE=ecc
```
```
#TPM1 requirements
#export CONFIG_TPM=y
```

> In the configuration, the path to the folder containing binary blobs must be specified. It is the line at the bottom of the board config `BOARD_TARGETS := NEW_BOARD` declaring the dependency. It corresponds to the name of the `mk file` referencing the needed binary blobs. These are required by coreboot for building the ROM.

Reviewer: ```suggestion
In the configuration, the binary blobs must be specified as build targets, as coreboot depends on these blobs. It is the line at the bottom of the board config `BOARD_TARGETS := NEW_BOARD`. It ensures that the main Heads makefile builds the targets specified in the `NEW_BOARD.mk` file. For more information on that file see below.
```

binary blobs:
---

> Check the coreboot configuration for the ported board. This will indicate which binary blobs are required and where they are expected to be located. Heads has architecture-/board-specific scripts under `blobs/*` which do the magic, called by the main makefile that handles everything in Heads. These scripts must create reproducible binary blobs when invoked. The blobs should be placed in the `blobs/NEW_BOARD/`. Make sure the coreboot config file references the correct path for each blob.

Reviewer: ```suggestion
Check the coreboot configuration for the ported board. This will indicate which binary blobs are required and where they are expected to be located. Heads expects architecture-/board-specific scripts under `blobs/*` which do the magic, called by the main makefile that handles everything in Heads. These scripts must create reproducible binary blobs when invoked. The blobs should be placed in the `blobs/NEW_BOARD/`. Make sure the coreboot config file references the correct path for each blob and the blobs are added to `.gitignore` so that they are not accidentally committed to git.
```

Generally,  three binary blobs are required: Management Engine (ME), Intel Flash Descriptor Region (IFD), and Gigabit Ethernet (GBE) The IFD and GBE can be extracted from a donor board using coreboot’s ifdtool. For more details, refer to the [upstream documentation](https://doc.coreboot.org/util/ifdtool/layout.html). On some boards it may be necessary to provide more blobs to coreboot, for instance an MRC blob for ram initialization if coreboot cannot distribute the blob itself for legal reasons.

Please note, if the ME is neutered, the IFD, coreboot CBFS region, and ME neutering space should be adjusted accordingly. Rationale: the ME region defined under IFD must fit. With the IFD ME region reduced, the BIOS region can grow with the freed ME space. With the BIOS region augmented, the CBFS region of the coreboot configuration must be increased to fit the ["maximized" space](https://osresearch.net/Prerequisites#legacy-vs-maximized-boards).

> Please note the GBE MAC address should be forged to: `00:DE:AD:C0:FF:EE MAC`. It can be done with [nvmutil](https://libreboot.org/docs/install/nvmutil.html). Due to licensing restrictions, the ME firmware cannot be uploaded to the GitHub. However, scripts can be used to build it locally and within CircleCI (a gray area legally, but still possible).

Reviewer: ```suggestion
Please note the GBE MAC address should be forged to: `00:DE:AD:C0:FF:EE MAC`. It can be done with [nvmutil](https://libreboot.org/docs/install/nvmutil.html). Due to licensing restrictions, the ME firmware cannot be uploaded to the GitHub. However, scripts can be used to build it locally and within CircleCI (a gray area legally, but still possible) GBE and IFD blobs can be uploaded directly to GitHub. Please explain how they were obtained in the commit message..
```

* Note: When calling scripts in Nix-based environments, Python must be invoked explicitly, as Nix does not allow executing Python scripts directly from files. One can use last clean example for t480: `python ./finalimage.py` will work and just `./finalimage.py` will not work. 
The blobs folder should have a script.sh which handles downloading, deactivating ME etc. It should also contain a README.md file briefly explaining the process. Hashes of the blobs should be stored either in `README.md` or in `hashes.txt file`. Furthermore, the script must ensure the integrity of the blobs it produced by comparing a SHA256 hash.

NEW_BOARD.mk:
---
Create a new `heads/targets/NEW_BOARD.mk` file which deals with calling blobs/script.sh*, download and extraction, and placing the blobs in correct location. This will be used by the main makefile and defines the board specific targets and dependencies, such as the binary blobs. For additional details, please see [Makefiles]({{ site.baseurl }}/Development/make-details.md).

coreboot.config:
---

> You need a coreboot configuration for the new board. Ideally, this config should be adapted from the closest possible platform. You can inspect existing configurations for boards in the master branch and select one with a similar architecture, if available. Next, you can create a coreboot config by copying it from closest platform and adapt it accordingly. You may use coreboot's `make menuconfig` for `NEW_BOARD`:

Reviewer: ```suggestion
You need a coreboot configuration for the new board. Ideally, this config should be adapted from the closest possible platform. You can inspect existing configurations for boards in the master branch and select one with a similar architecture, if available. Next, you can create a coreboot config by copying it from the closest platform and adapt it accordingly. You may use coreboot's `make menuconfig` for `NEW_BOARD`:
```

TPM measured boot (```CONFIG_TPM_MEASURED_BOOT=y```) and verified boot (```CONFIG_VBOOT_LIB=y```) should be enabled. Ensure that you select CONFIG_TPM_MEASURED_BOOT and CONFIG_VBOOT_LIB in Kconfig. Furthermore, enable the TPM and pick the correct TPM version for the board. 
Other parameters depend on the board. It is up to you to determine the correct settings, as not all community members will have access to your board. `git diff` between `coreboot-CLOSEST_PLATFORM.config` and `coreboot-NEW_BOARD.config` may help.

linux.config:
---
This should be adapted from a similar board. The file path should be `heads/config/linux-NEW_BOARD.config`.

CircleCI: 
---
Modify `heads/.circleci/config.yml` to add support for the `NEW_BOARD`. Initially, configure it to depend directly on musl-cross-make. At this stage, do not reuse caches—this simplifies debugging and ensures a clean build process. 
```
# NEW_BOARD is based on `X.Y.Z`. coreboot release, not sharing any buildstack from now, depend on muscl-cross cache
      - build:
          name: NEW_BOARD-hotp-maximized
          target: NEW_BOARD-hotp-maximized
          subcommand: ""
          requires:
            - x86-musl-cross-make
```

> CircleCI creates reproducible builds, allowing users to verify that the "flashable" roms were indeed produced by the given source code. In the development/debugging phase, it helps to ensure that you talk about the same build. It optimizes the collaboration between peers-board owners. Moreover, CI builds are significantly faster than local builds, reducing overall development time. As circleCI always builds all boards it makes it easier to find regression early.

Reviewer: ```suggestion
CircleCI creates reproducible builds, allowing users to verify that the "flashable" roms were indeed produced by the given source code. In the development/debugging phase, it helps to ensure that you talk about the same build. It optimizes the collaboration between peers-board owners. Moreover, CI builds are significantly faster than local builds, reducing overall development time. As circleCI always builds all boards it makes it easier to find regressions early.
```

nav_order: 3
parent: Development
---

<!-- markdownlint-disable MD033 -->
<details open markdown="block">
  <summary>
    Table of contents
  </summary>
  {: .text-delta }
1. TOC
{:toc}
</details>
<!-- markdownlint-enable MD033 -->

Prerequisites for porting Heads
===
Prerequisites: 

> 1. coreboot port:  since Heads is a coreboot payload, the board must have fully completed and actively maintained coreboot support. Any known issues should be acceptable to end users. Exceptions include the Librem and Dasharo coreboot forks, where coreboot is used from their respective forks.

Reviewer 2: Exceptions include the Purism and Dasharo coreboot forks, where coreboot is used from their respective forks.

4. External programmer: an external programmer is required to flash Heads and, if necessary, to recover from a brick.
5. Users and Board-testers: There should be multiple users and testers interested in using the board, as this makes maintenance and testing easier and benefits the entire ecosystem. You may, of course, port it for yourself first and allow others to join later. However, you must commit to testing—especially after coreboot version bumps or Linux kernel updates. If testing is not done in a timely manner, the board will be moved to an "unmaintained, untested" status. This would be unfortunate, potentially a waste of time, and disappointing for everyone involved.
6. GPU: dGPUs are problematic in Heads for various reasons. While successful ports on older machines with dGPUs exist (TODO: Add link), security may be affected, and there is currently no clear solution for that. (TODO: Link to Matrix coreboot discussion)

The scheme depicts a port cycle. 
![Port cycle]({{ site.baseurl }}/images/Porting_cycle.png)
TODO (remove source code after feedback)
```mermaid
  graph TD;
      build_rom-->flash
      flash-->test;

>       test-->debugg;

Reviewer 2: debugg -> debug

      debugg-->test;
      debugg-->remove_debugging_mode;
      remove_debugging_mode-->write_flashing/dissassambly_guide_on_heads-wiki;
      write_flashing/dissassambly_guide_on_heads-wiki-->maintance
```
Files needed to be created/modified are depicted on the scheme.
![Files overview]({{ site.baseurl }}/images/Files_scheme.png)

TODO (remove source code after feedback)
```mermaid
  graph LR;
      heads/-->config/;
      config/-->coreboot-new_board.config;
      config/-->linux-new_board.config;
      heads/-->boards/new_board-hotp-maximized/;
      heads/-->boards/new_board-maximized/;
      boards/new_board-hotp-maximized/-->new_board-hotp-maximized.config;
      boards/new_board-maximized/-->new_board-maximized.config;
      heads/-->blobs/new_board/;
      blobs/new_board/-->ME.bin;
      blobs/new_board/-->IFD.bin;
      blobs/new_board/-->GBE.bin;
      heads/-->targets/;
      targets/-->new_board.mk
      heads/-->modules/;
      modules/-->coreboot
      heads/-->.circleci/;
      .circleci/-->config.yml
```

Prerequisites for building the ROM
===
The goal is to build a ROM before the code review. `new_board` refers to the name of your board, e.g., `t480`

> coreboot.config:

Reviewer 2: Not sure how to add IFD <-> coreboot CBFS region <-> ME neutering freed space transfer here.

Maybe add a TODO where you see fit.

Rationale:
- If ME neutered, ME region defined under IFD must fit.
- With IFD ME region reduced, BIOS region can grow with ME freed space
- With BIOS region augmented, CBFS region of coreboot config must be augmented to fit "maximized" space 

Ref: https://osresearch.net/Prerequisites#legacy-vs-maximized-boards

---
you need a coreboot configuration. Ideally, this config should have already been tested and confirmed to work on the board. Alternatively, you can create one using coreboot's `make menuconfig`.
To ensure everything functions as expected, you may first build the ROM with SeaBIOS and test it. Please follow the upstream documentation, as this will save you time.

> Next, the configuration needs to be adapted for Heads, specifying LinuxBoot as the payload and defining the path for required blobs. You can inspect existing configurations for boards in the master branch and select one with a similar architecture (if available). The configuration file should be placed at: `heads/config/coreboot-new_board.config`

Reviewer 2: Unfortunately, that won't work, but should. Didn't tested: would build linuxboot payload with u-root, I'm not aware this was ever tested here. Hence why my previous comment.

Reviewer: I think bzImage is a requirement for the payload here:
```
CONFIG_PAYLOAD_LINUX=y
CONFIG_PAYLOAD_FILE="@BOARD_BUILD_DIR@/bzImage"
CONFIG_LINUX_INITRD="@BOARD_BUILD_DIR@/initrd.cpio.xz"
```

Generally,  three binary blobs are required: Management Engine (ME), Intel Flash Descriptor Region (IFD), and Gigabit Ethernet (GBE) The IFD and GBE can be extracted from a donor board using coreboot’s ifdtool. For more details, refer to the [uptsream documentation](https://doc.coreboot.org/util/ifdtool/layout.html)
Please note the GBE MAC address should be forged to: `00:DE:AD:C0:FF:EE MAC`. It can be done with [nvmutil](https://libreboot.org/docs/install/nvmutil.html). Due to licensing restrictions, the ME firmware cannot be uploaded to GitHub. However, scripts can be used to build it locally and within CircleCI (a gray area legally, but still possible).
* Note: When calling scripts in Nix-based environments, Python must be invoked explicitly, as Nix does not allow executing Python scripts directly from files. One can use last clean example for t480:`python ./finalimage.py` will work and just `./finalimage.py` will not work. 
Blobs folder should have a script.sh which handles downloading, deactivating ME etc. It should also contain README.md file briefly explaining the process. Hashes of the blobs should stored either in `README.md` or in `hashes.txt file`.

new_board.mk:
---

> create a new `targets/new_board.mk` file which deals with calling blobs/script.sh* download and extraction of blobs placing the blobs in correct location.

Reviewer: ```suggestion
Create a new `targets/new_board.mk` file. This will be used by the main makefile and defines the board specific targets and dependencies, such as the binary blobs.
Please not that you should add a line at the bottom of the board config `BOARD_TARGETS := new_board` to declare the dependency.
```

CircleCI: 
---

> modify `.circleci/config.yml` to add support for `new_board`. Initially, configure it to depend directly on musl-cross-make. At this stage, do not reuse caches—this simplifies debugging and ensures a clean build process. 

Reviewer: ```suggestion
Modify `.circleci/config.yml` to add support for `new_board`. Initially, configure it to depend directly on musl-cross-make. At this stage, do not reuse caches—this simplifies debugging and ensures a clean build process. 
```

```
# new_board is based on xx.xx coreboot release, not sharing any buildstack from now, depend on muscl-cross cache
      - build:
          name: new_board-hotp-maximized
          target: new_board-hotp-maximized
          subcommand: ""
          requires:
            - x86-musl-cross-make
```

> Under under CircleCI builds are reproducible, meaning you will be able to use the same (bit-to-bit) ROM on different machines to make sure you talk about the same thing, with same flashing technique/regions therefore SPI content. It optimizes the collaboration between peers-board owners. Moreover, CI builds are significantly faster than local builds, reducing overall development time.

Reviewer: ```suggestion
CircleCI creates reproducible builds, allowing users to verify that the flashable roms were indeed produced by the given source code. In the development/debugging phase, it helps to ensure that you talk about the same build. It optimizes the collaboration between peers-board owners. Moreover, CI builds are significantly faster than local builds, reducing overall development time.
```

The last part is only true when building all boards or maybe when doing it cleanly locally. As circleci always builds all boards it makes it easier to find regression early.

* Optional: local builds.

> If you really need to build locally during the development stage, pay attention to the helper functions at the end of the Makefiles. It is strongly recommended that local builders review the end of the Makefiles (including modules/* files), as these helper functions were designed to facilitate coreboot and Linux version bumps.

Reviewer: I would rephrase it. I think it makes a lot of sense to build locally when you are not a 100% sure something works. And it is actually faster if you only rebuild what you need to. It will also save some ci build time. So I would not discourage people as much from doing that.

	
Testing
===

For thorough testing—especially for a new board—using the following template may be beneficial.
Tasks to check, Copy and paste this template into the GitHub message window when reporting test results. Mark completed tasks with an x inside the brackets [ ]. Replace yyy with the relevant information
```
- [ ] Successful external flash link to circleci: https://yyy from commit `yyy` using external programmer model `yyy` on `yyy` Voltage mode
- [ ] Boots successfully after the flashing: 
- [ ] Setting clock prompt on first reboot: ok if triggered correctly after initial flashing and cmos battery disconnected
- [ ] Clean boot detected (no keyring, nothing installed on disk): usb boot proposed and followed: ok
- [ ] Boots on usb: ok
- [ ] OS `yyy` install and reboot: ok
- [ ] Heads functionality- no pubkey detected, but OS detected -> OEM-Factory-reset proposed. Done with `yyy` hardwarekey e.g. nk3: ok
- [ ] On reboot after re-ownership: generate new HOTP/TOTP: ok
- [ ] On reboot: default boot proposes to choose boot default + TPM DUK: ok
- [ ] TPM DUK boots OS version: ok

> - [ ] TPM DUK boots QuebesOS 4.2.4: ok

Reviewer 2: Please remove any reference to specific versions. Please use X.Y.Z where needed for posterity of the docs, here and anywhere else in the docs (coreboot version before as well being 24.12)

Author: done

- [ ] wifi works based on OS `yyy`: ok
- [ ] PR0: ok. 
* flashprog -p internal (not locked)

> * lock_chip (locks_

Reviewer 2: lock_chip (locks the platform with PR0, if PR0 patch applied in fork or under patches/coreboot-X.Y.Z, applied, and coreboot config applying proper preparation of the platform so lock_chip can simply lock the platform thorough PR0

See prior work
- https://github.com/linuxboot/heads/pull/1373
- https://github.com/linuxboot/heads/pull/1818
- unmerged patchset upstream https://review.coreboot.org/c/coreboot/+/85278

* flashprog -p internal (reports locked)
```

Debugging
===
Future debugging from Heads: enter recovery console, with a formatted USB thumb drive (fat32/exfat/ext3/ext4) and then 
```shell
mount-usb --mode rw
cp /tmp/debug.log /media

> cbmem -L > /media/coreboot_measured_boot.log

Reviewer 2: cbmem -L > /media/coreboot_measured_boot.txt
cbmem -1 > /media/coreboot_cbmem_console.txt
cbmem -t > /media/coreboot_cbmem_stages_timestamps.txt

umount /media
```
And then provide log files in a subsequent comment. That would be really helpful if something is still wrong. Of course, this should be modified based on the problem faced.

Pull Request
===

Create an initial PR, with commits relative roughly to the steps above. Please see the guide for contributing. 1 commit does 1 major bullet point.

> Then lets collaborate in the PR. Create a heads fork in your own repository, follow/comment linuxboot/heads-wiki#119 TODO create the wiki-page as you go, make sure you sign commits and don't stress too much if this seems a lot. Get a PR started and we will collaborate there. 

Reviewer: ```suggestion
Then lets collaborate in the PR. Create a heads fork in your own repository, follow/comment linuxboot/heads-wiki#119 TODO create the wiki-page as you go, make sure you sign and sign-off commits and don't stress too much if this seems a lot. Get a PR started and we will collaborate there. 
```
